### PR TITLE
In standalone mode, ensure cached tokens are valid before returning them

### PR DESCRIPTION
**Pull request by zarenner**

### CODE REVIEW

**Comment by zarenner** [comment]
Still need to verify against public feeds

File: CredentialProvider.Microsoft/Util/SessionTokenCache.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@ public SessionTokenCache(string cacheFilePath, ILogger logger, CancellationToken
         {
             this.cacheFilePath = cacheFilePath;
             this.logger = logger;
+            this.cancellationToken = cancellationToken;
```
Comment (zarenner) [inline CredentialProvider.Microsoft/Util/SessionTokenCache.cs]: Looked like this was missing. There's still oddness though in that the request handler interface in nuget requires a token both as a member and as a parameter. I think we're getting default tokens in some places we shouldn't - worth looking at sometime.

File: CredentialProvider.Microsoft/CredentialProviderArgs.cs [hunk 2]
```diff
@@ -19,9 +19,14 @@ internal class CredentialProviderArgs
...
         public bool NonInteractive { get; set; }
 
-        [ArgDescription("Notifies the provider that this is a retry and the credentials were rejected on a previous attempt")]
+        [ArgDescription("Always generate fresh credentials, ignoring the cache")]
+        [OmitFromUsageDocs] // IsRetry support is really for the NuGet flow and is confusing for standalone. Keep for backcompat, but don't advertise.
         public bool IsRetry { get; set; }
 
+        [ArgDescription("Skip validating cached credentials. Expired or revoked credentials may be returned.")]
+        [ArgShortcut("S")]
+        public bool SkipValidatingCachedCreds { get; set; }
```
Comment (zarenner) [inline CredentialProvider.Microsoft/CredentialProviderArgs.cs]: Thoughts on not even adding this option and just always validating? Even if someone is already using the IsRetry flow, the one extra GET isn't really significant overhead.

Comment (salvmd) [reply CredentialProvider.Microsoft/CredentialProviderArgs.cs]: I agree to nuke it

File: CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs [hunk 3]
```diff
@@ -8,47 +8,89 @@
...
 using NuGetCredentialProvider.CredentialProviders;
-using NuGetCredentialProvider.CredentialProviders.Vsts;
 using NuGetCredentialProvider.Logging;
 using NuGetCredentialProvider.Util;
 
 namespace NuGetCredentialProvider.RequestHandlers
 {
     /// <summary>
-    /// Handles a <see cref="GetAuthenticationCredentialsRequest"/> and replies with credentials.
+    /// Handles a <see cref="ExtendedGetAuthenticationCredentialsRequest"/> and replies with credentials.
```
Comment (salvmd) [inline CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs]: update comment

File: CredentialProvider.Microsoft/Program.cs [hunk 4]
```diff
@@ -167,7 +171,7 @@ public static async Task<int> Main(string[] args)
                     }
 
                     GetAuthenticationCredentialsRequest request = new GetAuthenticationCredentialsRequest(parsedArgs.Uri, isRetry: parsedArgs.IsRetry, isNonInteractive: parsedArgs.NonInteractive, parsedArgs.CanShowDialog);
-                    GetAuthenticationCredentialsResponse response = await getAuthenticationCredentialsRequestHandler.HandleRequestAsync(request).ConfigureAwait(continueOnCapturedContext: false);
```
Comment (salvmd) [inline CredentialProvider.Microsoft/Program.cs]: Nuke getAuthenticationCredentialsRequestHandler if not used anymore

File: CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs [hunk 5]
```diff
@@ -8,47 +8,89 @@
...
+        public GetAuthenticationCredentialsOrchestrator(
+            ICache<Uri, string> cache,
+            ILogger logger,
+            IReadOnlyCollection<ICredentialProvider> credentialProviders,
+            ICachedCredentialsValidator credentialsValidator)
         {
-            this.credentialProviders = credentialProviders ?? throw new ArgumentNullException(nameof(credentialProviders));
             this.cache = cache;
+            this.logger = logger;
+            this.credentialProviders = credentialProviders;
```
Comment (salvmd) [inline CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs]: Add null checks back

File: CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs [hunk 6]
```diff
@@ -8,47 +8,89 @@
...
         private readonly IReadOnlyCollection<ICredentialProvider> credentialProviders;
-        private readonly TimeSpan progressReporterTimeSpan = TimeSpan.FromSeconds(2);
+        private readonly ICachedCredentialsValidator credentialsValidator;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="GetAuthenticationCredentialsRequestHandler"/> class.
         /// </summary>
         /// <param name="logger">A <see cref="ILogger"/> to use for logging.</param>
         /// <param name="credentialProviders">An <see cref="IReadOnlyCollection{ICredentialProviders}"/> containing credential providers.</param>
         /// <param name="cache">An <see cref="ICache{TKey, TValue}"/> cache to store found credentials.</param>
```
Comment (salvmd) [inline CredentialProvider.Microsoft/RequestHandlers/GetAuthenticationCredentialsRequestHandler.cs]: nit: add credentialsValidator here

**Comment by zarenner** [comment]
Abandoning this PR - we've decided to continue having the caller of the cred provider be responsible for indicating that a cached token no longer works (and retrying).

**Pull request closed without merging**